File: app/src/main/java/com/berkatfaatulohalawa1711010164/facevoting/API/APIService.java
package com.berkatfaatulohalawa1711010164.facevoting.API;

import android.content.Context;

import com.berkatfaatulohalawa1711010164.facevoting.config.Constants;
import com.berkatfaatulohalawa1711010164.facevoting.model.AkunModel;
import com.berkatfaatulohalawa1711010164.facevoting.model.LoginModel;
import com.berkatfaatulohalawa1711010164.facevoting.model.MenuModel;
import com.berkatfaatulohalawa1711010164.facevoting.model.MessageModel;
import com.berkatfaatulohalawa1711010164.facevoting.model.PaslonModel;
import com.berkatfaatulohalawa1711010164.facevoting.model.UserModel;
import com.berkatfaatulohalawa1711010164.facevoting.response.GetMenu;
import com.berkatfaatulohalawa1711010164.facevoting.response.GetPaslon;
import com.berkatfaatulohalawa1711010164.facevoting.response.GetPerolehan;
import com.berkatfaatulohalawa1711010164.facevoting.response.GetVote;

import java.util.concurrent.TimeUnit;

import okhttp3.Cache;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface APIService {
    @GET("api/Suara/perolehan")
    Call<GetPerolehan> tampilSuara(@Query("id_kategori") String id_kategori);

    @GET("api/Kategori/semua")
    Call<GetMenu> tampilHasil();


    @GET("api/Suara/tampil/{id_user}")
    Call<GetVote> tampilVote(@Path("id_user") String id_user);

    @FormUrlEncoded
    @PUT("api/Akun/tampil/{id_user}")
    Call<AkunModel> updateAkun(@Path("id_user") String id_user,
                               @Field("nama") String nama,
                               @Field("identitas") String identitas,
                               @Field("password") String password);

    @GET("api/Akun/tampil")
    Call<AkunModel> tampilAKun(@Query("id_user") String id_user);



    @FormUrlEncoded
    @POST("api/Suara/vote")
    Call<MessageModel> simpanVote(@Field("id_user") String id_user,
                                  @Field("id_kategori") String id_kategori,
                                  @Field("id_paslon") String id_paslon);

    @Multipart
    @POST("api/Gambar/cek")
    Call<MessageModel> cekWajah(@Part("id_user") RequestBody id,
                                  @Part MultipartBody.Part upload);
    @Multipart
    @POST("api/Gambar/tambah")
    Call<MessageModel> rekamWajah(@Part("id_user") RequestBody id,
                                  @Part MultipartBody.Part upload);


    @FormUrlEncoded
    @POST("api/Login/sudahlogin")
    Call<LoginModel> dapatstatus(@Field("id_user") String id_user);

    @FormUrlEncoded
    @POST("api/Login/otentikasi")
    Call<LoginModel> validasiLogin(@Field("email") String email,
                               @Field("password") String password);

    //API untuk menambah data mahasiswa
    @FormUrlEncoded
    @POST("api/Daftar/tambah")
    Call<UserModel> daftarUser(@Field("nama") String nama,
                                    @Field("identitas") String identitas,
                                    @Field("email") String email,
                                    @Field("password") String password,
                                    @Field("token_firebase") String token_firebase);

    //Mendapatkan data paslon berdasarkan id kategori / ukm atau orkem
    @FormUrlEncoded
    @POST("api/Paslon/tampil")
    Call<GetPaslon> postPaslon(@Field("id_kategori") String id_kategori);

    //API untuk menampilkan kategori daftar pemilihan umum
    @GET("api/Kategori/tampil")
    Call<GetMenu> listMenu(@Query("id_user") String id_user);

    class Factory{
        public static APIService create(Context mContext){
            OkHttpClient.Builder builder = new OkHttpClient.Builder();

            builder.readTimeout(20, TimeUnit.SECONDS);
            builder.connectTimeout(20, TimeUnit.SECONDS);
            builder.writeTimeout(20, TimeUnit.SECONDS);
            builder.addInterceptor(new NetworkConnectionInterceptor(mContext));

            OkHttpClient client = builder.build();

            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl(Constants.URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();

            return retrofit.create(APIService.class);
        }
    }
}

File: android/facevoting/app/src/main/java/com/berkatfaatulohalawa1711010164/facevoting/API/APIService.kt
package com.berkatfaatulohalawa1711010164.facevoting.api

import android.content.Context
import com.berkatfaatulohalawa1711010164.facevoting.config.Constants
import com.berkatfaatulohalawa1711010164.facevoting.model.AkunModel
import com.berkatfaatulohalawa1711010164.facevoting.model.LoginModel
import com.berkatfaatulohalawa1711010164.facevoting.model.MessageModel
import com.berkatfaatulohalawa1711010164.facevoting.model.UserModel
import com.berkatfaatulohalawa1711010164.facevoting.response.GetMenu
import com.berkatfaatulohalawa1711010164.facevoting.response.GetPaslon
import com.berkatfaatulohalawa1711010164.facevoting.response.GetPerolehan
import com.berkatfaatulohalawa1711010164.facevoting.response.GetVote
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query
import java.util.concurrent.TimeUnit

interface APIService {
    @GET("api/Login/cekstatus")
    fun cekStatus(@Query("id_user") idUser: String): Call<LoginModel>

    @GET("api/Suara/perolehan")
    fun tampilSuara(@Query("id_kategori") idKategori: String): Call<GetPerolehan>

    @GET("api/Kategori/semua")
    fun tampilHasil(): Call<GetMenu>

    @GET("api/Suara/tampil/{id_user}")
    fun tampilVote(@Path("id_user") idUser: String): Call<GetVote>

    @FormUrlEncoded
    @PUT("api/Akun/tampil/{id_user}")
    fun updateAkun(
        @Path("id_user") idUser: String,
        @Field("nama") nama: String,
        @Field("identitas") identitas: String,
        @Field("password") password: String
    ): Call<AkunModel>

    @GET("api/Akun/tampil")
    fun tampilAkun(@Query("id_user") idUser: String): Call<AkunModel>

    @FormUrlEncoded
    @POST("api/Suara/vote")
    fun simpanVote(
        @Field("id_user") idUser: String,
        @Field("id_kategori") idKategori: String,
        @Field("id_paslon") idPaslon: String
    ): Call<MessageModel>

    @Multipart
    @POST("api/Gambar/cek")
    fun cekWajah(
        @Part("id_user") id: RequestBody,
        @Part upload: MultipartBody.Part
    ): Call<MessageModel>

    @Multipart
    @POST("api/Gambar/tambah")
    fun rekamWajah(
        @Part("id_user") id: RequestBody,
        @Part upload: MultipartBody.Part
    ): Call<MessageModel>

    @FormUrlEncoded
    @POST("api/Login/sudahlogin")
    fun dapatStatus(@Field("id_user") idUser: String): Call<LoginModel>

    @FormUrlEncoded
    @POST("api/Login/otentikasi")
    fun validasiLogin(
        @Field("email") email: String,
        @Field("password") password: String
    ): Call<LoginModel>

    @FormUrlEncoded
    @POST("api/Daftar/tambah")
    fun daftarUser(
        @Field("nama") nama: String,
        @Field("identitas") identitas: String,
        @Field("email") email: String,
        @Field("password") password: String,
        @Field("token_firebase") tokenFirebase: String
    ): Call<UserModel>

    @FormUrlEncoded
    @POST("api/Paslon/tampil")
    fun postPaslon(@Field("id_kategori") idKategori: String): Call<GetPaslon>

    @GET("api/Kategori/tampil")
    fun listMenu(@Query("id_user") idUser: String): Call<GetMenu>

    companion object {
        fun create(mContext: Context?): APIService {
            val builder = OkHttpClient.Builder()
                .readTimeout(20, TimeUnit.SECONDS)
                .connectTimeout(20, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .addInterceptor(NetworkConnectionInterceptor(mContext))

            val client = builder.build()

            val retrofit = Retrofit.Builder()
                .baseUrl(Constants.URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

            return retrofit.create(APIService::class.java)
        }
    }
}
